### PR TITLE
Add CI action to update lists weekly

### LINKED ISSUE #4: set up an Action to regularly generate new lists
For example, we could update the generated lists once per day if I add a github token as a CI secret.

### PR DESCRIPTION
**Pull request by khos2ow**
Generate lists (top 100, 1000, and 10000) weekly and push back changes to
the repository.

Fixes #4 

Signed-off-by: Khosrow Moossavi <khos2ow@gmail.com>

### CODE REVIEW

File: .github/workflows/daily.yml
Comment (khos2ow) [inline .github/workflows/daily.yml]: Git user and user-email can be configured in appropriate values. Note that these don't need to be actual valid username and emails.

**Review by khos2ow — COMMENTED** [review]
@mvdan check my comment below, and also note that `GITHUB_TOKEN` is being injected by GitHub Actions and no extra steps needed on your side.

File: .github/workflows/daily.yml
Comment (mvdan) [reply .github/workflows/daily.yml]: how about `auto-updater@users.noreply.github.com`?

**Review by mvdan — APPROVED** [review]
Thanks for this!

I think my initial idea to do this daily is overkill and will just cause noise. I think weekly makes a lot more sense, since the list won't really change much from one day to another. If you can make that adjustment, I'm happy to merge.

File: .github/workflows/daily.yml
Comment (mvdan) [inline .github/workflows/daily.yml]: also, these three steps are a bit wasteful. just generate the 10k list, then use `sed` to chunk off the top 1k and 100.

File: .github/workflows/daily.yml
Comment (mvdan) [inline .github/workflows/daily.yml]: 1.16.x?

File: .github/workflows/daily.yml
Comment (mvdan) [inline .github/workflows/daily.yml]: please pin a specific full commit here. I don't want to give another user the ability to do anything with my github token by just pushing to that repository's master branch and waiting a few days.

File: .github/workflows/daily.yml
Comment (cristaloleg) [reply .github/workflows/daily.yml]: `1.x` ?

Comment (mvdan) [reply .github/workflows/daily.yml]: I think that's a bit overkill :) I'd rather do major versions explicitly. Bugfix versions basically never break stuff, but major releases sometimes do.

Comment (khos2ow) [reply .github/workflows/daily.yml]: I didn't get this, do you mean to pin it on specific version or anything in 1.16 range? If the latter `go-version: 1.16` would be `>= 1.16, < 1.17`.

File: .github/workflows/daily.yml
Comment (khos2ow) [reply .github/workflows/daily.yml]: Great idea!

File: .github/workflows/daily.yml
Comment (khos2ow) [reply .github/workflows/daily.yml]: Pinned on [`13266270ee2da98f1d16425f6ce21c235a7f33dd`](https://github.com/drud/action-cross-commit/commits/13266270ee2da98f1d16425f6ce21c235a7f33dd)

File: .github/workflows/daily.yml
Comment (mvdan) [reply .github/workflows/daily.yml]: I mean literally the 1.16.x string, which will do the latest bugfix release.

Comment (khos2ow) [reply .github/workflows/daily.yml]: My understanding was that `1.16` is equivalent of `1.16.x`. I'll update regardless.

Comment (mvdan) [reply .github/workflows/daily.yml]: Maybe, though that's a bit confusing as Go calls the first release 1.16, not 1.16.0. The ".x" helps disambiguate.

**Review by mvdan — APPROVED** [review]

**Comment by mvdan** [comment]
@khos2ow hm, how did you test this? it appears to have failed: https://github.com/mvdan/corpus/runs/2541292629?check_suite_focus=true

Maybe I need to add a secret of some sort.

**Comment by khos2ow** [comment]
Oh I see what's happening. [`GITHUB_TOKEN`](https://docs.github.com/en/actions/reference/authentication-in-a-workflow#about-the-github_token-secret) is being injected by GitHub automatically but it is a secret and not an environment variable.

**Comment by khos2ow** [comment]
#8 fixes the issue.

**Comment by mvdan** [comment]
Unfortunately we trigger the abuse heuristic now: https://github.com/mvdan/corpus/runs/2596564898?check_suite_focus=true

It worked for me locally, so maybe it's because my personal token has higher rate limits than the default CI one. If it fails again next weekend, I'll try swapping it out.

**Comment by khos2ow** [comment]
Interesting, I just tried with a brand new generated token locally and got the exact same error. There has to be some other factors in play here. This [says](https://docs.github.com/en/rest/overview/resources-in-the-rest-api#rate-limiting):

> For API requests using Basic Authentication or OAuth, you can make up to 5,000 requests per hour.
> ...
> For users that belong to a GitHub Enterprise Cloud account, requests made using an OAuth token to resources owned by the same GitHub Enterprise Cloud account have an increased limit of 15,000 requests per hour.
> ...
> When using the built-in GITHUB_TOKEN in GitHub Actions, the rate limit is 1,000 requests per hour per repository

So if your token works locally I suspect you are part of an enteprise account, otherwise it seems more weird than it is!

Even 1,000 should be more than enough for the CI, and 5,000 should be way more than enough for my local test. For verbosity I tried outputing ratelimit using `vlogf` and:

```text
corpus on master ! took 8s ➜ GITHUB_TOKEN="xxxxxx" go run . -v -count 10000 > top-10000.ts
v
2021/05/17 13:20:01 querying first page of results for "archived:false is:public pushed:>=2020-05-18 language:go sort:stars" 
2021/05/17 13:20:01 1 API called
2021/05/17 13:20:07 Ratelimit, cost: 1, limit: 5000, remaining: 4923
...
2021/05/17 13:22:00 1748/10000 done; querying page 24 with cursor Y3Vyc29yOjMwMA==
2021/05/17 13:22:00 24 API called
2021/05/17 13:22:07 Ratelimit, cost: 1, limit: 5000, remaining: 4900
...
2021/05/17 13:22:07 1822/10000 done; querying page 25 with cursor Y3Vyc29yOjQwMA==
2021/05/17 13:22:07 25 API called
non-200 OK status code: 403 Forbidden body: "{\n  \"documentation_url\": \"https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#abuse-rate-limits\",\n  \"message\": \"You have triggered an abuse detection mechanism. Please wait a few minutes before you try again.\"\n}\n"
exit status 1
```

As you can see I started with 4923 remaining and it broke when, realistically, I still had 4900 remaining in my quota.

**Comment by khos2ow** [comment]
All that said, it might worth exploring the possibility to even breaking down process further away in the code to have more graceful API call (maybe with some `sleep` in between?) Technically there's only one API call happening in the code, but very frequent.

This also [says](https://github.com/google/go-github#rate-limiting):

> ... The Search API has a custom rate limit. Unauthenticated clients are limited to 10 requests per minute, while authenticated clients can make up to 30 requests per minute.

But then again based on [GraphQL rate limit](https://docs.github.com/en/graphql/overview/resource-limitations#rate-limit) our query is very flat and shouldn't cost _too much_.

**Comment by khos2ow** [comment]
@mvdan did you get a chance to try it with your personal token? Although I think it might be better to refactor a bit to more gracefully call GitHub APIs to make this work with any provided token. What do you think?

**Comment by mvdan** [comment]
Swapped the token here, we'll see if it helps: https://github.com/mvdan/corpus/commit/209c5a85555d8af657bbf5b9fd993b356104c9bf

If it doesn't help, I might just decrease the limit to unblock the updates. I could give you write access if you want to try other methods, too :) For now, I want to see what the next run in three days does.

**Comment by mvdan** [comment]
Nnnnnope: https://github.com/mvdan/corpus/runs/3216198836?check_suite_focus=true

@khos2ow do you want to attempt to fix this? If so, I can give you write access and you should be able to push to feature branches and trigger workflows.

**Comment by mvdan** [comment]
I've also opened https://github.com/mvdan/corpus/issues/9 to track this, since this PR is merged.

**Comment by khos2ow** [comment]
@mvdan yeah, I'm gonna take a crack at it.

**Pull request merged**
